MethodNotFoundException occurred while store token with redis

the exception was
```
Handling error: NestedServletException, Handler dispatch failed; nested exception is java.lang.NoSuchMethodError: org.springframework.data.redis.connection.RedisConnection.set([B[B)V
```

Env:
- Spring Boot M6
- Spring Data Redis 2.0.1.RELEASE
- Spring Cloud Finchley.M4

Code throws which throws exception blow:
line 157 `conn.set(accessKey, serializedAccessToken); `

https://github.com/spring-projects/spring-security-oauth/blob/0c034e4dfac53c6f736a3d7e858bae37328e0fab/spring-security-oauth2/src/main/java/org/springframework/security/oauth2/provider/token/store/redis/RedisTokenStore.java#L154-L163

I think this issue needs to be re-opened. As things stand org.springframework.security.oauth.boot:spring-security-oauth2-autoconfigure:2.0.0.RELEASE does not work with Spring Boot 2.0 when using Redis to store tokens.

As has already been noted above, the underlying problem is that `org.springframework.security.oauth:spring-security-oauth2` compiles against Spring Data Redis 1.5 whereas Boot 2.0 uses Spring Data Redis 2.0.x. The latter is not binary compatible with the former so a failure occurs at runtime:

```
Caused by: java.lang.NoSuchMethodError: org.springframework.data.redis.connection.RedisConnection.set([B[B)V
	at org.springframework.security.oauth2.provider.token.store.redis.RedisTokenStore.storeAccessToken(RedisTokenStore.java:160) ~[spring-security-oauth2-2.2.1.RELEASE.jar:na]
	at org.springframework.security.oauth2.provider.token.DefaultTokenServices.createAccessToken(DefaultTokenServices.java:122) ~[spring-security-oauth2-2.2.1.RELEASE.jar:na]
	at org.springframework.security.oauth2.provider.token.AbstractTokenGranter.getAccessToken(AbstractTokenGranter.java:70) ~[spring-security-oauth2-2.2.1.RELEASE.jar:na]
	at org.springframework.security.oauth2.provider.token.AbstractTokenGranter.grant(AbstractTokenGranter.java:65) ~[spring-security-oauth2-2.2.1.RELEASE.jar:na]
	at org.springframework.security.oauth2.provider.CompositeTokenGranter.grant(CompositeTokenGranter.java:38) ~[spring-security-oauth2-2.2.1.RELEASE.jar:na]
	at org.springframework.security.oauth2.config.annotation.web.configurers.AuthorizationServerEndpointsConfigurer$4.grant(AuthorizationServerEndpointsConfigurer.java:561) ~[spring-security-oauth2-2.2.1.RELEASE.jar:na]
	at org.springframework.security.oauth2.provider.endpoint.TokenEndpoint.postAccessToken(TokenEndpoint.java:132) ~[spring-security-oauth2-2.2.1.RELEASE.jar:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_151]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_151]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_151]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_151]
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:209) ~[spring-web-5.0.4.RELEASE.jar:5.0.4.RELEASE]
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:136) ~[spring-web-5.0.4.RELEASE.jar:5.0.4.RELEASE]
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:102) ~[spring-webmvc-5.0.4.RELEASE.jar:5.0.4.RELEASE]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:870) ~[spring-webmvc-5.0.4.RELEASE.jar:5.0.4.RELEASE]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:776) ~[spring-webmvc-5.0.4.RELEASE.jar:5.0.4.RELEASE]
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87) ~[spring-webmvc-5.0.4.RELEASE.jar:5.0.4.RELEASE]
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:991) ~[spring-webmvc-5.0.4.RELEASE.jar:5.0.4.RELEASE]
	... 76 common frames omitted
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -12,5 +12,5 @@ install: ./mvnw -U install --quiet -DskipTests=true -P bootstrap
 script:
   - jdk_switcher use openjdk7
   - ./mvnw clean test -P bootstrap
-#  - jdk_switcher use oraclejdk8
-#  - ./mvnw -U clean test -P spring5
+  - jdk_switcher use oraclejdk8
+  - ./mvnw -f spring-security-oauth2 -U clean test -P spring5

--- a/pom.xml
+++ b/pom.xml
@@ -21,6 +21,8 @@
     <commons-codec.version>1.9</commons-codec.version>
     <spring.version>4.0.9.RELEASE</spring.version>
     <spring.security.version>3.2.10.RELEASE</spring.security.version>
+    <spring.data.redis.version>1.5.0.RELEASE</spring.data.redis.version>
+    <redis.clients.version>2.6.3</redis.clients.version>
     <java.version>1.6</java.version>
   </properties>
 
@@ -165,8 +167,10 @@
     <profile>
       <id>spring5</id>
         <properties>
-            <spring.version>5.0.0.BUILD-SNAPSHOT</spring.version>
-            <spring.security.version>5.0.0.BUILD-SNAPSHOT</spring.security.version>
+            <spring.version>5.0.4.RELEASE</spring.version>
+            <spring.security.version>5.0.3.RELEASE</spring.security.version>
+            <spring.data.redis.version>2.0.5.RELEASE</spring.data.redis.version>
+            <redis.clients.version>2.9.0</redis.clients.version>
         </properties>
         <repositories>
             <repository>

--- a/spring-security-oauth2/pom.xml
+++ b/spring-security-oauth2/pom.xml
@@ -161,14 +161,14 @@
 		<dependency>
         	<groupId>org.springframework.data</groupId>
 			<artifactId>spring-data-redis</artifactId>
-			<version>1.5.0.RELEASE</version>
+			<version>${spring.data.redis.version}</version>
 			<optional>true</optional>
 		</dependency>
 
 		<dependency>
 			<groupId>redis.clients</groupId>
 			<artifactId>jedis</artifactId>
-			<version>2.6.3</version>
+			<version>${redis.clients.version}</version>
 			<optional>true</optional>
 		</dependency>
 

--- a/spring-security-oauth2/src/main/java/org/springframework/security/oauth2/client/http/OAuth2ErrorHandler.java
+++ b/spring-security-oauth2/src/main/java/org/springframework/security/oauth2/client/http/OAuth2ErrorHandler.java
@@ -118,7 +118,7 @@ public class OAuth2ErrorHandler implements ResponseErrorHandler {
 				}
 
 				public int getRawStatusCode() throws IOException {
-					return response.getRawStatusCode();
+					return this.getStatusCode().value();
 				}
 			};
 

--- a/spring-security-oauth2/src/main/java/org/springframework/security/oauth2/provider/token/store/redis/RedisTokenStore.java
+++ b/spring-security-oauth2/src/main/java/org/springframework/security/oauth2/provider/token/store/redis/RedisTokenStore.java
@@ -1,11 +1,5 @@
 package org.springframework.security.oauth2.provider.token.store.redis;
 
-import java.util.ArrayList;
-import java.util.Collection;
-import java.util.Collections;
-import java.util.Date;
-import java.util.List;
-
 import org.springframework.data.redis.connection.RedisConnection;
 import org.springframework.data.redis.connection.RedisConnectionFactory;
 import org.springframework.security.oauth2.common.ExpiringOAuth2RefreshToken;
@@ -15,6 +9,15 @@ import org.springframework.security.oauth2.provider.OAuth2Authentication;
 import org.springframework.security.oauth2.provider.token.AuthenticationKeyGenerator;
 import org.springframework.security.oauth2.provider.token.DefaultAuthenticationKeyGenerator;
 import org.springframework.security.oauth2.provider.token.TokenStore;
+import org.springframework.util.ClassUtils;
+import org.springframework.util.ReflectionUtils;
+
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Date;
+import java.util.List;
 
 /**
  * @author efenderbosch
@@ -31,14 +34,23 @@ public class RedisTokenStore implements TokenStore {
 	private static final String CLIENT_ID_TO_ACCESS = "client_id_to_access:";
 	private static final String UNAME_TO_ACCESS = "uname_to_access:";
 
+	private static final boolean springDataRedis_2_0 = ClassUtils.isPresent(
+			"org.springframework.data.redis.connection.RedisStandaloneConfiguration",
+			RedisTokenStore.class.getClassLoader());
+
 	private final RedisConnectionFactory connectionFactory;
 	private AuthenticationKeyGenerator authenticationKeyGenerator = new DefaultAuthenticationKeyGenerator();
 	private RedisTokenStoreSerializationStrategy serializationStrategy = new JdkSerializationStrategy();
 	
 	private String prefix = "";
 
+	private Method redisConnectionSet_2_0;
+
 	public RedisTokenStore(RedisConnectionFactory connectionFactory) {
 		this.connectionFactory = connectionFactory;
+		if (springDataRedis_2_0) {
+			this.loadRedisConnectionMethods_2_0();
+		}
 	}
 
 	public void setAuthenticationKeyGenerator(AuthenticationKeyGenerator authenticationKeyGenerator) {
@@ -51,6 +63,11 @@ public class RedisTokenStore implements TokenStore {
 	
 	public void setPrefix(String prefix) {
 		this.prefix = prefix;
+	}
+
+	private void loadRedisConnectionMethods_2_0() {
+		this.redisConnectionSet_2_0 = ReflectionUtils.findMethod(
+				RedisConnection.class, "set", byte[].class, byte[].class);
 	}
 
 	private RedisConnection getConnection() {
@@ -157,9 +174,19 @@ public class RedisTokenStore implements TokenStore {
 		RedisConnection conn = getConnection();
 		try {
 			conn.openPipeline();
-			conn.set(accessKey, serializedAccessToken);
-			conn.set(authKey, serializedAuth);
-			conn.set(authToAccessKey, serializedAccessToken);
+			if (springDataRedis_2_0) {
+				try {
+					this.redisConnectionSet_2_0.invoke(conn, accessKey, serializedAccessToken);
+					this.redisConnectionSet_2_0.invoke(conn, authKey, serializedAuth);
+					this.redisConnectionSet_2_0.invoke(conn, authToAccessKey, serializedAccessToken);
+				} catch (Exception ex) {
+					throw new RuntimeException(ex);
+				}
+			} else {
+				conn.set(accessKey, serializedAccessToken);
+				conn.set(authKey, serializedAuth);
+				conn.set(authToAccessKey, serializedAccessToken);
+			}
 			if (!authentication.isClientOnly()) {
 				conn.rPush(approvalKey, serializedAccessToken);
 			}
@@ -177,9 +204,18 @@ public class RedisTokenStore implements TokenStore {
 				byte[] refresh = serialize(token.getRefreshToken().getValue());
 				byte[] auth = serialize(token.getValue());
 				byte[] refreshToAccessKey = serializeKey(REFRESH_TO_ACCESS + token.getRefreshToken().getValue());
-				conn.set(refreshToAccessKey, auth);
 				byte[] accessToRefreshKey = serializeKey(ACCESS_TO_REFRESH + token.getValue());
-				conn.set(accessToRefreshKey, refresh);
+				if (springDataRedis_2_0) {
+					try {
+						this.redisConnectionSet_2_0.invoke(conn, refreshToAccessKey, auth);
+						this.redisConnectionSet_2_0.invoke(conn, accessToRefreshKey, refresh);
+					} catch (Exception ex) {
+						throw new RuntimeException(ex);
+					}
+				} else {
+					conn.set(refreshToAccessKey, auth);
+					conn.set(accessToRefreshKey, refresh);
+				}
 				if (refreshToken instanceof ExpiringOAuth2RefreshToken) {
 					ExpiringOAuth2RefreshToken expiringRefreshToken = (ExpiringOAuth2RefreshToken) refreshToken;
 					Date expiration = expiringRefreshToken.getExpiration();

--- a/spring-security-oauth2/src/test/java/org/springframework/security/oauth2/config/annotation/Gh808EnableAuthorizationServerTests.java
+++ b/spring-security-oauth2/src/test/java/org/springframework/security/oauth2/config/annotation/Gh808EnableAuthorizationServerTests.java
@@ -33,6 +33,8 @@ import org.springframework.security.core.userdetails.UserDetails;
 import org.springframework.security.core.userdetails.UserDetailsService;
 import org.springframework.security.core.userdetails.UsernameNotFoundException;
 import org.springframework.security.crypto.codec.Base64;
+import org.springframework.security.crypto.password.NoOpPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
 import org.springframework.security.oauth2.config.annotation.configurers.ClientDetailsServiceConfigurer;
 import org.springframework.security.oauth2.config.annotation.web.configuration.AuthorizationServerConfigurerAdapter;
 import org.springframework.security.oauth2.config.annotation.web.configuration.EnableAuthorizationServer;
@@ -187,6 +189,11 @@ public class Gh808EnableAuthorizationServerTests {
 					return user;
 				}
 			};
+		}
+
+		@Bean
+		public PasswordEncoder passwordEncoder() {
+			return NoOpPasswordEncoder.getInstance();
 		}
 	}
 }

--- a/spring-security-oauth2/src/test/java/org/springframework/security/oauth2/config/annotation/ResourceServerConfigurationTests.java
+++ b/spring-security-oauth2/src/test/java/org/springframework/security/oauth2/config/annotation/ResourceServerConfigurationTests.java
@@ -36,6 +36,8 @@ import org.springframework.security.config.annotation.web.configuration.EnableWe
 import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
 import org.springframework.security.core.Authentication;
 import org.springframework.security.crypto.codec.Base64;
+import org.springframework.security.crypto.password.NoOpPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
 import org.springframework.security.oauth2.common.DefaultOAuth2AccessToken;
 import org.springframework.security.oauth2.common.OAuth2AccessToken;
 import org.springframework.security.oauth2.config.annotation.configurers.ClientDetailsServiceConfigurer;
@@ -253,6 +255,10 @@ public class ResourceServerConfigurationTests {
 
 		@Configuration
 		protected static class SecurityConfiguration extends WebSecurityConfigurerAdapter {
+			@Bean
+			public PasswordEncoder passwordEncoder() {
+				return NoOpPasswordEncoder.getInstance();
+			}
 		}
 	}
 
@@ -275,6 +281,10 @@ public class ResourceServerConfigurationTests {
 		}
 		@Configuration
 		protected static class SecurityConfiguration extends WebSecurityConfigurerAdapter {
+			@Bean
+			public PasswordEncoder passwordEncoder() {
+				return NoOpPasswordEncoder.getInstance();
+			}
 		}
 	}
 

--- a/spring-security-oauth2/src/test/java/org/springframework/security/oauth2/provider/token/store/redis/RedisTokenStoreTests.java
+++ b/spring-security-oauth2/src/test/java/org/springframework/security/oauth2/provider/token/store/redis/RedisTokenStoreTests.java
@@ -4,12 +4,17 @@ import org.junit.Before;
 import org.junit.Test;
 import org.springframework.data.redis.connection.jedis.JedisConnectionFactory;
 import org.springframework.security.authentication.TestingAuthenticationToken;
-import org.springframework.security.oauth2.common.*;
+import org.springframework.security.oauth2.common.DefaultExpiringOAuth2RefreshToken;
+import org.springframework.security.oauth2.common.DefaultOAuth2AccessToken;
+import org.springframework.security.oauth2.common.DefaultOAuth2RefreshToken;
+import org.springframework.security.oauth2.common.OAuth2AccessToken;
+import org.springframework.security.oauth2.common.OAuth2RefreshToken;
 import org.springframework.security.oauth2.provider.OAuth2Authentication;
 import org.springframework.security.oauth2.provider.OAuth2Request;
 import org.springframework.security.oauth2.provider.RequestTokenFactory;
 import org.springframework.security.oauth2.provider.token.TokenStore;
 import org.springframework.security.oauth2.provider.token.store.TokenStoreBaseTests;
+import org.springframework.util.ClassUtils;
 import redis.clients.jedis.JedisShardInfo;
 
 import java.util.Collection;
@@ -32,8 +37,18 @@ public class RedisTokenStoreTests extends TokenStoreBaseTests {
 
 	@Before
 	public void setup() throws Exception {
-		JedisShardInfo shardInfo = new JedisShardInfo("localhost");
-		JedisConnectionFactory connectionFactory = new JedisConnectionFactory(shardInfo);
+		boolean springDataRedis_2_0 = ClassUtils.isPresent(
+				"org.springframework.data.redis.connection.RedisStandaloneConfiguration",
+				this.getClass().getClassLoader());
+
+		JedisConnectionFactory connectionFactory;
+		if (springDataRedis_2_0) {
+			connectionFactory = new JedisConnectionFactory();
+		} else {
+			JedisShardInfo shardInfo = new JedisShardInfo("localhost");
+			connectionFactory = new JedisConnectionFactory(shardInfo);
+		}
+
 		tokenStore = new RedisTokenStore(connectionFactory);
 	}
 

--- a/spring-security-oauth2/src/test/resources/org/springframework/security/oauth2/config/xml/authorization-server-client-credentials-password-invalid.xml
+++ b/spring-security-oauth2/src/test/resources/org/springframework/security/oauth2/config/xml/authorization-server-client-credentials-password-invalid.xml
@@ -52,6 +52,7 @@
 
 	<authentication-manager id="usersAuthenticationManager">
 		<authentication-provider>
+			<password-encoder ref="passwordEncoder" />
 			<user-service>
 				<user name="acme" password="password" authorities="ROLE_USER" />
 			</user-service>
@@ -64,5 +65,7 @@
 		<oauth2:client-credentials />
 		<oauth2:password authentication-manager-ref="usersAuthenticationManager"  />
 	</oauth2:authorization-server>
+
+	<b:bean id="passwordEncoder" class="org.springframework.security.crypto.password.NoOpPasswordEncoder" factory-method="getInstance" />
 
 </b:beans>

--- a/spring-security-oauth2/src/test/resources/org/springframework/security/oauth2/config/xml/authorization-server-client-credentials-password-valid.xml
+++ b/spring-security-oauth2/src/test/resources/org/springframework/security/oauth2/config/xml/authorization-server-client-credentials-password-valid.xml
@@ -33,7 +33,9 @@
 	</b:bean>
 
 	<authentication-manager id="clientsAuthenticationManager">
-		<authentication-provider user-service-ref="clientDetailsUserDetailsService" />
+		<authentication-provider user-service-ref="clientDetailsUserDetailsService">
+			<password-encoder ref="passwordEncoder" />
+		</authentication-provider>
 	</authentication-manager>
 
 	<b:bean id="clientDetailsUserDetailsService" class="org.springframework.security.oauth2.provider.client.ClientDetailsUserDetailsService">
@@ -46,6 +48,7 @@
 
 	<authentication-manager id="usersAuthenticationManager">
 		<authentication-provider>
+			<password-encoder ref="passwordEncoder" />
 			<user-service>
 				<user name="acme" password="password" authorities="ROLE_USER" />
 			</user-service>
@@ -58,5 +61,7 @@
 		<oauth2:client-credentials />
 		<oauth2:password authentication-manager-ref="usersAuthenticationManager"  />
 	</oauth2:authorization-server>
+
+	<b:bean id="passwordEncoder" class="org.springframework.security.crypto.password.NoOpPasswordEncoder" factory-method="getInstance" />
 
 </b:beans>

